From e658d57ac90e344f84322369e41f449876830411 Mon Sep 17 00:00:00 2001
From: myfluxi <linflux@arcor.de>
Date: Wed, 27 Nov 2013 17:07:03 +0100
Subject: [PATCH 133/590] msm: mdss: Add safety check for user input

Leave more checks either to the control app or the user's intelligence.

Signed-off-by: Francisco Franco <franciscofranco.1990@gmail.com>
Signed-off-by: Simarpreet Singh <simar@linux.com>
---
 drivers/video/msm/mdss/mdss_dsi_panel.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/video/msm/mdss/mdss_dsi_panel.c b/drivers/video/msm/mdss/mdss_dsi_panel.c
index b9f11a7..3882f94 100644
--- a/drivers/video/msm/mdss/mdss_dsi_panel.c
+++ b/drivers/video/msm/mdss/mdss_dsi_panel.c
@@ -1092,11 +1092,12 @@ static int write_local_on_cmds(struct device *dev, const char *buf,
 			       size_t cmd)
 {
 	int i, rc = 0;
-	int val;
+	unsigned int val;
 	int dlen, offset = 0;
 	bool rgb;
 	char tmp[3];
 	struct mdss_dsi_ctrl_pdata *ctrl = NULL;
+	struct mdss_panel_common_pdata *prev_local_data;
 
 	if (cnt) {
 		cnt = 0;
@@ -1119,11 +1120,19 @@ static int write_local_on_cmds(struct device *dev, const char *buf,
 	if (!dlen)
 		return -EINVAL;
 
+	prev_local_data = local_pdata;
+
 	for (i = offset; i < dlen; i++) {
 		rc = sscanf(buf, "%d", &val);
 		if (rc != 1)
 			return -EINVAL;
 
+		if (val > 255) {
+			pr_err("%s: Invalid input data %u (0-255)\n", __func__, val);
+			local_pdata = prev_local_data;
+			return -EINVAL;
+		}
+
 		local_pdata->on_cmds.cmds[cmd].payload[i] = val;
 		if (rgb)
 			local_pdata->on_cmds.cmds[cmd + 2].payload[i] = val;
-- 
1.7.9.5

